File: fapi/src/main/java/com/netcracker/edu/fapi/service/PollService.java
package com.netcracker.edu.fapi.service;

import com.netcracker.edu.fapi.models.viewModels.ClonePoll;
import com.netcracker.edu.fapi.models.viewModels.ViewPoll;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;

import java.util.List;


public interface PollService {

    ViewPoll findById(int id);

    List<ViewPoll> findAllTemplateByTheme(String theme);

    ResponseEntity<?> save(ViewPoll poll);

    void deletePoll(int id);

    ViewPoll findByLink(String link);

    ViewPoll clonePoll(ClonePoll clonePoll);

    ViewPoll submitPoll(int id);

    Page<ViewPoll> findAllByUserId(int userId, String theme, String substr,
                                   int page, int size, String sort, String order);

    Page<ViewPoll> findDraftsByUserId(int idUser, String theme, String substr,
                                      int page, int size, String sort, String order);

    Page<ViewPoll> findActivePollsByUserId(int idUser, String theme, String substr,
                                           int page, int size, String sort, String order);

    Page<ViewPoll> getPolls(int userId, String select, String theme, String substr,
                            int page, int size, String sort, String order);
}

File: frontend/src/app/services/poll.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {PollModel} from "../modules/models/poll.model";
import {ClonePollModel} from "../modules/models/clonePoll.model";
import {RestPageModel} from "../modules/models/rest-page.model";


@Injectable()
export class PollService {
  private page: RestPageModel;
  private _currPoll: PollModel;

  get currPoll(): PollModel {
    return this._currPoll;
  }

  set currPoll(value: PollModel) {
    this._currPoll = value;
  }


  constructor(private http: HttpClient) {
  }

  getPollsPageByUserId(userId: number, select: string, theme: string, search: string,
                       page: number, size: number, sort: string, order: string): Observable<RestPageModel> {
    return this.http.get<RestPageModel>('/api/poll/user' + '?userId=' + userId + '&select=' + select + '&theme=' +
      theme + '&substr=' + search + '&page=' + page + '&size=' + size + '&sort=' + sort + '&order=' + order);
  }

  getPollById(id: string) {
    return this.http.get<PollModel>('/api/poll/' + id);
  }

  savePoll(poll: PollModel): Observable<PollModel> {
    return this.http.post<PollModel>('/api/poll', poll);
  }

  deletePoll(id: string): Observable<void> {
    return this.http.delete<void>('/api/poll/' + id);
  }

  getAllTemplatesByTheme(theme: string): Observable<PollModel[]> {
    return this.http.get<PollModel[]>('/api/poll/template?theme=' + theme);
  }

  clonePoll(clonePoll: ClonePollModel): Observable<PollModel> {
    return this.http.post<PollModel>('/api/poll/clone', clonePoll);
  }

  submitPoll(id: number): Observable<PollModel> {
    return this.http.post<PollModel>('api/poll/submit/' + id, id);
  }

  getPollByLink(pollLink: string): Observable<PollModel> {
    return this.http.get<PollModel>('api/poll?link=' + pollLink);

  }
}
